src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -315,13 +316,50 @@ public async Task<object> GetDescriptionAsync(IAsyncCompletionSession session, V
             }
         }
 
+        /// <summary>
+        /// We'd like to cache VS Completion item dircetly to avoid allocation completely. However it holds references
+        /// to transient objects, which would cause memory leak (among other potential issues) if cached. 
+        /// So as a compromise,  we cache data that can be calculated from Roslyn completion item to avoid repeated 
+        /// calculation cost for cached Roslyn completion items.
+        /// </summary>
+        private class VSCompletionItemData
+        {
+            public VSCompletionItemData(string displayText, ImageElement icon, ImmutableArray<AsyncCompletionData.CompletionFilter> filters, ImmutableArray<ImageElement> attributeIcons, string insertionText)
+            {
+                DisplayText = displayText;
+                Icon = icon;
+                Filters = filters;
+                AttributeIcons = attributeIcons;
+                InsertionText = insertionText;
+            }
+
+            public string DisplayText { get; }
+
+            public ImageElement Icon { get; }
+
+            public ImmutableArray<AsyncCompletionData.CompletionFilter> Filters { get; }
+
+            public ImmutableArray<ImageElement> AttributeIcons { get; }
+
+            public string InsertionText { get; }
+        }
+
         private VSCompletionItem Convert(
             Document document,
             RoslynCompletionItem roslynItem)
         {
-            if (roslynItem.IsCached && s_roslynItemToVsItem.TryGetValue(roslynItem, out var vsItem))
+            if (roslynItem.IsCached && s_roslynItemToVsItemData.TryGetValue(roslynItem, out var itemData))
             {
-                return vsItem;
+                return new VSCompletionItem(
```

Comment:
I wish this was refactored so we didn't have two places calling this constructor. It'd be nice if this just instead was a check of if we _missed_ the cache then it assigns itemData the values created below, and then outside the if we just consume roslynItem and itemData always so there's only a single constructor. The extra allocations would be a problem, but if we made VSCompletionItem a struct (and then used a `StrongBox<>` to hold it in the CWT), we could avoid that.

Reply:
Something like this?

```cs


        private VSCompletionItem Convert(
            Document document,
            RoslynCompletionItem roslynItem)
        {
            VSCompletionItemData itemData;

            if (roslynItem.IsCached && s_roslynItemToVsItemData.TryGetValue(roslynItem, out var boxedItemData))
            {
                itemData = boxedItemData.Value;
            }
            else
            {
                var imageId = roslynItem.Tags.GetFirstGlyph().GetImageId();
                var filters = GetFilters(roslynItem);

                // roslynItem generated by providers can contain an insertionText in a property bag.
                // We will not use it but other providers may need it.
                // We actually will calculate the insertion text once again when called TryCommit.
                if (!roslynItem.Properties.TryGetValue(InsertionText, out var insertionText))
                {
                    insertionText = roslynItem.DisplayText;
                }

                var supportedPlatforms = SymbolCompletionItem.GetSupportedPlatforms(roslynItem, document.Project.Solution.Workspace);
                var attributeImages = supportedPlatforms != null ? s_WarningImageAttributeImagesArray : ImmutableArray<ImageElement>.Empty;

                itemData = new VSCompletionItemData(
                    displayText: roslynItem.GetEntireDisplayText(),
                    icon: new ImageElement(new ImageId(imageId.Guid, imageId.Id), roslynItem.DisplayText),
                    filters: filters,
                    attributeIcons: attributeImages,
                    insertionText: insertionText);

                // It doesn't make sense to cache VS item data for those Roslyn items created from scratch for each session,
                // since CWT uses object identity for comparison.
                if (roslynItem.IsCached)
                {
                    s_roslynItemToVsItemData.Add(roslynItem, new StrongBox<VSCompletionItemData>(itemData));
                }
            }

            var item = new VSCompletionItem(
                displayText: itemData.DisplayText,
                source: this,
                icon: itemData.Icon,
                filters: itemData.Filters,
                suffix: roslynItem.InlineDescription, // InlineDescription will be right-aligned in the selection popup
                insertText: itemData.InsertionText,
                sortText: roslynItem.SortText,
                filterText: roslynItem.FilterText,
                attributeIcons: itemData.AttributeIcons);

            item.Properties.AddProperty(RoslynItem, roslynItem);

            return item;
        }
```